<div style="
  text-align: center; 
  margin: 40px 0;
  background: linear-gradient(135deg, #28a745, #00c853);
  padding: 20px 0;
  border-radius: 16px;
  box-shadow: 0 6px 14px rgba(0,0,0,0.15);
  transition: transform 0.3s ease, box-shadow 0.3s ease, background 0.3s ease;
  cursor: default;
" 
onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 10px 24px rgba(0,0,0,0.25)'; this.style.background='linear-gradient(135deg, #00c853, #28a745)'"
onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 6px 14px rgba(0,0,0,0.15)'; this.style.background='linear-gradient(135deg, #28a745, #00c853)'"
>
  <span style="
    font-family: 'Playfair Display', serif;
    color: black;
    font-size: 1.8em;
    font-weight: 700;
    font-style: italic;
    letter-spacing: 0.5px;
  ">
    SmartMart Sales Analysis
  </span>
</div>
<!-- <a id="import-data"></a> -->


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color:hsla(176, 94%, 54%, 0.91);
    color: black;
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
    Import Libraries
  </span>
</div>
<a id="import-data"></a>


In [43]:

import pandas as pd
import numpy as np
import plotly.graph_objects as go
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.figure_factory as ff
from datetime import datetime, timedelta, date
from operator import attrgetter
import warnings
warnings.filterwarnings('ignore')

# Enhanced Plotly configuration for premium visuals
import plotly.io as pio
pio.templates.default = "plotly_white"
pio.renderers.default = "notebook_connected"
# Configuration
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color:hsla(176, 94%, 54%, 0.91);
    color: black;
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
    Import Dataset
  </span>
</div>
<a id="import-data"></a>


In [44]:
path = r'dataset.xlsx' # path to the dataset
xls = pd.ExcelFile(path) 

sheet_names = [s for s in xls.sheet_names if s != "Customers"]

Customers = pd.read_excel(path, sheet_name="Customers")

Customers = pd.merge(
    Customers,
    pd.read_excel(path, sheet_name="Regions"),
    left_on="Region_ID",
    right_on="Region_ID",
    how="left"
)
master = pd.read_excel(path, sheet_name="Orders")


FK_IDs = ['Product_ID', 'Sales_Rep_ID', 'Region_ID']

for i in range(len(FK_IDs)):
    detail = pd.read_excel(path, sheet_name=sheet_names[i+1])  # +1 لأن Orders هو master
    master = pd.merge(master, detail, left_on=FK_IDs[i], right_on=FK_IDs[i], how="left")

master = pd.merge(
    master,
    Customers[['Customer_ID', 'Customer_Name' ,'Region_ID',]],
    left_on="Region_ID",
    right_on="Region_ID",
    how="left"
)

In [45]:
df = master.copy()

In [46]:
df.head()

Order_ID Order_Date  Product_ID  Sales_Rep_ID  Region_ID  Quantity_Sold  Unit_Price  Total_Sales     Profit  Return_Flag   Time  Year  Month Day_of_Week  Hour Product_Name Product_Category   Sales_Rep_Name  Hire_Date  Region_Name  Customer_ID       Customer_Name
0      1001 2024-09-05         202           305        101              2     523.336     1046.672  198.86768        False  23:47  2024      9    Thursday    23  Televisions      Electronics  Dragoș Petrescu 2023-05-19  Târgu Mureș          402   Ana-Maria Ionescu
1      1001 2024-09-05         202           305        101              2     523.336     1046.672  198.86768        False  23:47  2024      9    Thursday    23  Televisions      Electronics  Dragoș Petrescu 2023-05-19  Târgu Mureș          407       Vlad Petrescu
2      1001 2024-09-05         202           305        101              2     523.336     1046.672  198.86768        False  23:47  2024      9    Thursday    23  Televisions      Electronics  Dragoș Petrescu 2023-05-19  Târgu Mureș          410   Georgiana Ionescu
3      1001 2024-09-05         202           305        101              2     523.336     1046.672  198.86768        False  23:47  2024      9    Thursday    23  Televisions      Electronics  Dragoș Petrescu 2023-05-19  Târgu Mureș          419     Ștefan Petrescu
4      1001 2024-09-05         202           305        101              2     523.336     1046.672  198.86768        False  23:47  2024      9    Thursday    23  Televisions      Electronics  Dragoș Petrescu 2023-05-19  Târgu Mureș          420  Cristina Georgescu

In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40113 entries, 0 to 40112
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Order_ID          40113 non-null  int64         
 1   Order_Date        40113 non-null  datetime64[ns]
 2   Product_ID        40113 non-null  int64         
 3   Sales_Rep_ID      40113 non-null  int64         
 4   Region_ID         40113 non-null  int64         
 5   Quantity_Sold     40113 non-null  int64         
 6   Unit_Price        40113 non-null  float64       
 7   Total_Sales       40113 non-null  float64       
 8   Profit            40113 non-null  float64       
 9   Return_Flag       40113 non-null  bool          
 10  Time              40113 non-null  object        
 11  Year              40113 non-null  int64         
 12  Month             40113 non-null  int64         
 13  Day_of_Week       40113 non-null  object        
 14  Hour              4011

<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color:hsla(176, 94%, 54%, 0.91);
    color: black;
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
    Check for missing and duplicate values
  </span>
</div>
<a id="import-data"></a>


In [48]:
print("Missing values percentage: \n" ,  (df.isna().sum() / len(df)) * 100)

Missing values percentage: 
 Order_ID            0.0
Order_Date          0.0
Product_ID          0.0
Sales_Rep_ID        0.0
Region_ID           0.0
Quantity_Sold       0.0
Unit_Price          0.0
Total_Sales         0.0
Profit              0.0
Return_Flag         0.0
Time                0.0
Year                0.0
Month               0.0
Day_of_Week         0.0
Hour                0.0
Product_Name        0.0
Product_Category    0.0
Sales_Rep_Name      0.0
Hire_Date           0.0
Region_Name         0.0
Customer_ID         0.0
Customer_Name       0.0
dtype: float64


In [49]:
print("Duplicate rows:  " , df.duplicated().sum())

Duplicate rows:   0


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color:hsla(176, 94%, 54%, 0.91);
    color: black;
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
     Descriptive Statistics
  </span>
</div>
<a id="import-data"></a>


In [50]:
df_descriptive = df.drop(columns =['Order_ID', 'Product_ID', 'Sales_Rep_ID', 'Region_ID', 'Customer_ID','Year' , 'Month'])
df_descriptive.describe(include=[float, int]).round(2)

Quantity_Sold  Unit_Price  Total_Sales    Profit      Hour
count       40113.00    40113.00     40113.00  40113.00  40113.00
mean            5.42      351.70      1904.00    404.24     11.57
std             2.84      179.36      1483.56    340.53      6.89
min             1.00       44.11        44.11      8.51      0.00
25%             3.00      193.75       693.70    144.17      6.00
50%             5.00      351.41      1480.78    300.37     12.00
75%             8.00      505.34      2808.04    585.24     18.00
max            10.00      659.91      6583.28   1974.98     23.00

In [51]:
df_descriptive.describe(include=[object, "category"]).T


count unique               top   freq
Time              40113   1083             21:12    138
Day_of_Week       40113      7          Thursday   6342
Product_Name      40113      5       Televisions   8797
Product_Category  40113      3       Electronics  24206
Sales_Rep_Name    40113      5  Cristian Popescu   8568
Region_Name       40113      5            Brăila  10738
Customer_Name     40113     53      Ioan Popescu   1216

<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
     1. What is the month-over-month (MoM) and year-over-year (YoY) sales growth rate?
  </span>
</div>
<a id="import-data"></a>


In [52]:

df['Order_Date'] = pd.to_datetime(df['Order_Date'])
df['Period'] = df['Order_Date'].dt.to_period('M')


monthly = (
    df
    .groupby('Period', as_index=True)
    .agg(Total_Sales=('Total_Sales', 'sum'))
    .sort_index()
    .reset_index()
)
monthly['PeriodStart'] = monthly['Period'].dt.to_timestamp()
monthly['MoM_pct'] = monthly['Total_Sales'].pct_change(periods=1) * 100
monthly['YoY_pct'] = monthly['Total_Sales'].pct_change(periods=12) * 100

fig = make_subplots(specs=[[{"secondary_y": True}]])

fig.add_trace(
    go.Bar(
        x=monthly['PeriodStart'],
        y=monthly['Total_Sales'],
        name='Total Sales',
        marker_color='rgba(58, 71, 80, 0.8)',
        hovertemplate='%{x|%Y-%m}<br>Total Sales: %{y:,.2f}$<extra></extra>'
    ),
    secondary_y=False
)

# خط MoM
fig.add_trace(
    go.Scatter(
        x=monthly['PeriodStart'],
        y=monthly['MoM_pct'],
        mode='lines+markers',
        name='MoM %',
        line=dict(color='orange', width=3),
        marker=dict(size=8),
        hovertemplate='%{x|%Y-%m}<br>MoM Growth: %{y:.2f}%<extra></extra>'
    ),
    secondary_y=True
)

fig.add_trace(
    go.Scatter(
        x=monthly['PeriodStart'],
        y=monthly['YoY_pct'],
        mode='lines+markers',
        name='YoY %',
        line=dict(color='red', width=3, dash='dash'),
        marker=dict(size=8),
        hovertemplate='%{x|%Y-%m}<br>YoY Growth: %{y:.2f}%<extra></extra>'
    ),
    secondary_y=True
)

fig.add_hline(y=0, line=dict(color='black', width=1, dash='dot'), secondary_y=True)

fig.update_layout(
 title=dict(
    text='Monthly Sales with MoM and YoY Growth (%)',
        x=0.5,                
        xanchor='center',
        font=dict(size=30)
    ),
    xaxis=dict(title='Month'),
    yaxis=dict(title='Total Sales', showgrid=True),
    yaxis2=dict(title='Growth (%)', showgrid=False),
    legend=dict(orientation='h', yanchor='bottom', y=1.02, xanchor='right', x=1),
    hovermode='x unified',
    template='plotly_dark',
    bargap=0.2,
    margin=dict(l=60, r=60, t=80, b=60)
    
)

fig.update_yaxes(tickformat=".1f", secondary_y=True)
fig.show()



In [53]:
monthly.round(2)


Period  Total_Sales PeriodStart  MoM_pct  YoY_pct
0   2022-10    646821.71  2022-10-01      NaN      NaN
1   2022-11   3153180.98  2022-11-01   387.49      NaN
2   2022-12   4133705.31  2022-12-01    31.10      NaN
3   2023-01   3074404.15  2023-01-01   -25.63      NaN
4   2023-02   2586119.38  2023-02-01   -15.88      NaN
5   2023-03   3197590.70  2023-03-01    23.64      NaN
6   2023-04   2849979.55  2023-04-01   -10.87      NaN
7   2023-05   3271113.82  2023-05-01    14.78      NaN
8   2023-06   3334550.93  2023-06-01     1.94      NaN
9   2023-07   3060160.07  2023-07-01    -8.23      NaN
10  2023-08   3288875.90  2023-08-01     7.47      NaN
11  2023-09   3574832.84  2023-09-01     8.69      NaN
12  2023-10   2751196.71  2023-10-01   -23.04   325.34
13  2023-11   3243330.40  2023-11-01    17.89     2.86
14  2023-12   3689237.38  2023-12-01    13.75   -10.75
15  2024-01   2749441.55  2024-01-01   -25.47   -10.57
16  2024-02   3253291.82  2024-02-01    18.33    25.80
17  2024-03   2398017.03  2024-03-01   -26.29   -25.01
18  2024-04   3236138.27  2024-04-01    34.95    13.55
19  2024-05   3416214.34  2024-05-01     5.56     4.44
20  2024-06   3770445.01  2024-06-01    10.37    13.07
21  2024-07   3107876.71  2024-07-01   -17.57     1.56
22  2024-08   3263641.24  2024-08-01     5.01    -0.77
23  2024-09   2584285.02  2024-09-01   -20.82   -27.71
24  2024-10   2740594.54  2024-10-01     6.05    -0.39

# Sales Insights (Oct 2022 – Oct 2024)

---

## 1. Initial Growth Phase (Oct–Dec 2022)
- Sales jumped sharply from **646K (Oct 2022)** to **4.13M (Dec 2022)**.
- **Nov 2022** saw the strongest monthly growth (**+387% MoM**) — likely due to a **seasonal event (promotions, holidays, or launch activity)**.

---

## 2. Fluctuations in 2023
- After peaking in **Dec 2022**, sales **dropped 25% in Jan 2023** and **16% in Feb 2023**.
- 2023 showed **alternating growth and decline cycles**, suggesting sensitivity to **campaigns or external factors**.
- **Sep 2023** hit **3.57M (+8.7% MoM)**, a strong month.
- **Oct 2023** dropped -23% MoM but showed **+325% YoY vs Oct 2022**.
- Overall: **volatile but upward trend compared to 2022**.

---

## 3. Performance in 2024
- **Jan 2024:** Weak start (**-25.5% MoM, -10.6% YoY**).
- **Feb 2024:** Strong rebound (**+18% MoM, +25.8% YoY**).
- **Mar 2024:** Sharp decline (**-26% MoM, -25% YoY**) → one of the weakest months.
- **Apr–Jun 2024:** Strong recovery  
  - Apr: **+35% MoM, +13.5% YoY**  
  - Jun: **3.77M (+10% MoM, +13% YoY)** → one of the best months in 2024.
- **Jul–Sep 2024:** Downturn  
  - Jul: **-17.6% MoM, +1.6% YoY**  
  - Sep: **-20.8% MoM, -27.7% YoY** → major concern.
- **Oct 2024:** Slight recovery (**+6% MoM**) but **flat YoY (-0.4%)**.

---

## 4. Key Patterns
- **Seasonality:** Strong **Nov–Dec peaks** (holiday/end-of-year).
- **Q1 Weakness:** Jan–Mar generally weaker (seen in 2023 & 2024).
- **Mid-Year Recovery:** Apr–Jun rebounds strongly.
- **Volatility:** Sales swings heavily → driven by **campaigns/events**.
- **2024 Warning Signs:** H1 was strong, but **Q3 saw steep YoY declines**.

---

## 5. Actionable Insights
- **Maximize Q4 campaigns** (Nov–Dec) to capture peak growth.
- **Reduce Q1 slumps** with **new campaigns or retention strategies**.
- **Investigate Q3 2024 drop** → possible causes:  
  - Market saturation  
  - Competitor activity  
  - Reduced promotions
- **Stabilize growth** → use **loyalty programs, consistent offers, and new channels** to reduce reliance on seasonal spikes.


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
    2. What is the average order size?
  </span>
</div>
<a id="import-data"></a>


In [54]:

order_summary = df.groupby('Order_ID').agg({
    'Total_Sales':'sum',
    'Quantity_Sold':'sum'
}).reset_index()

order_summary['Average_Order_Size'] = order_summary['Total_Sales'] / order_summary['Quantity_Sold']

mean_order_size = order_summary['Average_Order_Size'].mean()
median_order_size = order_summary['Average_Order_Size'].median()

print(f"Average Order Size (Mean): ${mean_order_size:.2f}")
print(f"Average Order Size (Median): ${median_order_size:.2f}")


Average Order Size (Mean): $352.42
Average Order Size (Median): $351.47


In [55]:


region_summary = df.groupby('Region_Name').agg({
    'Total_Sales':'sum',
    'Quantity_Sold':'sum'
}).reset_index()
region_summary['Average_Order_Size'] = region_summary['Total_Sales'] / region_summary['Quantity_Sold']

fig = px.bar(
    region_summary.round(2),
    x='Region_Name',
    y='Average_Order_Size',
    # text='Average_Order_Size',           
    labels={'Average_Order_Size':'Average Order Size ($)', 'Region_Name':'Region'},
    color_discrete_sequence=['#34ebd8']
)

fig.update_layout(
    title='Average Order Size per Region',
    title_x=0.5,
    template='plotly_dark'
)

region_summary

Region_Name   Total_Sales  Quantity_Sold  Average_Order_Size
0       Brăila  2.055021e+07          59046          348.037346
1      Craiova  1.584351e+07          44562          355.538465
2       Galați  1.277580e+07          37485          340.824249
3     Ploiești  1.245514e+07          34663          359.320767
4  Târgu Mureș  1.475039e+07          41515          355.302769

In [56]:

fig.show()

In [57]:
df['Order_Date'] = pd.to_datetime(df['Order_Date'])
monthly_summary = df.groupby(df['Order_Date'].dt.to_period('M')).agg({
    'Total_Sales':'sum',
    'Quantity_Sold':'sum'
}).reset_index().round(2)
monthly_summary['Average_Order_Size'] = monthly_summary['Total_Sales'] / monthly_summary['Quantity_Sold']
monthly_summary['Period'] = monthly_summary['Order_Date'].dt.strftime('%b %Y')

fig = px.line(
    monthly_summary.round(2),
    x='Period',
    y='Average_Order_Size',
    markers=True,
    text='Average_Order_Size',
    labels={'Average_Order_Size':'Average Order Size ($)', 'Period':'Month'}
)

fig.update_layout(
    title='Monthly Average Order Size',
    title_x=0.5,
    template='plotly_dark',
    xaxis_tickangle=-45
)

monthly_summary

Order_Date  Total_Sales  Quantity_Sold  Average_Order_Size    Period
0     2022-10    646821.71           1424          454.228729  Oct 2022
1     2022-11   3153180.98           8918          353.574902  Nov 2022
2     2022-12   4133705.31          10700          386.327599  Dec 2022
3     2023-01   3074404.15           9187          334.647235  Jan 2023
4     2023-02   2586119.38           7983          323.953323  Feb 2023
5     2023-03   3197590.70           8630          370.520359  Mar 2023
6     2023-04   2849979.55           7437          383.216290  Apr 2023
7     2023-05   3271113.82           8894          367.788826  May 2023
8     2023-06   3334550.93           9154          364.272551  Jun 2023
9     2023-07   3060160.07           7810          391.825873  Jul 2023
10    2023-08   3288875.90           9139          359.872623  Aug 2023
11    2023-09   3574832.84          10260          348.424253  Sep 2023
12    2023-10   2751196.71           8646          318.204570  Oct 2023
13    2023-11   3243330.40           9657          335.852791  Nov 2023
14    2023-12   3689237.38          10724          344.016913  Dec 2023
15    2024-01   2749441.55           8672          317.048149  Jan 2024
16    2024-02   3253291.82           9194          353.849447  Feb 2024
17    2024-03   2398017.03           7461          321.406920  Mar 2024
18    2024-04   3236138.27           8191          395.084638  Apr 2024
19    2024-05   3416214.34          10088          338.641390  May 2024
20    2024-06   3770445.01          10786          349.568423  Jun 2024
21    2024-07   3107876.71           8641          359.666324  Jul 2024
22    2024-08   3263641.24           9262          352.368953  Aug 2024
23    2024-09   2584285.02           8355          309.309996  Sep 2024
24    2024-10   2740594.54           8058          340.108531  Oct 2024

In [58]:
fig.show()


# Order Size & Regional Sales Insights

---

## 1. Overall Order Size
- **Average Order Size (Mean):** $352.42  
- **Average Order Size (Median):** $351.47  

👉 The mean and median are very close, indicating a **balanced distribution** of order sizes (no major skew from extreme outliers).

---

## 2. Regional Performance 
| Region       | Total Sales   | Quantity Sold | Avg. Order Size |
|--------------|--------------:|--------------:|----------------:|
| Brăila       | 20.55M        | 59,046        | $348.04 |
| Craiova      | 15.84M        | 44,562        | $355.54 |
| Galați       | 12.78M        | 37,485        | $340.82 |
| Ploiești     | 12.46M        | 34,663        | $359.32 |
| Târgu Mureș  | 14.75M        | 41,515        | $355.30 |

### Insights:
- **Brăila** leads in **total sales (20.5M)** and quantity sold.  
- **Ploiești** achieves the **highest average order size ($359.32)** among the top regions.  
- Most regions cluster around **$340–$360 per order**, showing consistency across geographies.

---

## 3. Order Size Trends Over Time
| Period  | Total Sales   | Orders | Avg. Order Size |
|---------|--------------:|-------:|----------------:|
| Oct 2022 | 0.65M | 1,424 | $454.23 |
| Nov 2022 | 3.15M | 8,918 | $353.57 |
| Dec 2022 | 4.13M | 10,700 | $386.33 |
| Jan 2023 | 3.07M | 9,187 | $334.65 |
| Feb 2023 | 2.59M | 7,983 | $323.95 |
| Mar 2023 | 3.20M | 8,630 | $370.52 |
| ... | ... | ... | ... |
| Sep 2024 | 2.58M | 8,355 | $309.31 |
| Oct 2024 | 2.74M | 8,058 | $340.11 |

### Key Observations:
- **High Order Value at Start:** Oct 2022 had the highest recorded order size (**$454**) — early stage, fewer orders, but high-value purchases.  
- **Stabilization:** From **2023 onwards**, average order sizes stabilized in the **$320–$380** range.  
- **Seasonality:**  
  - Stronger order sizes around **Nov–Dec (holiday peaks)**.  
  - **Drops in Jan–Feb** (weaker purchasing power after holidays).  
- **Recent Trend (2024):**  
  - Apr 2024 had a high **$395/order**, despite lower sales volume.  
  - Sep 2024 fell sharply to **$309/order** (lowest point), indicating **smaller basket sizes** or **discount-driven sales**.

---

## 4. Actionable Insights
- **Maintain High Basket Values:** Replicate strategies from Oct 2022 & Apr 2024 when order sizes peaked.  
- **Boost Q1:** Order size shrinks in **Jan–Feb each year** → consider **bundles, loyalty discounts, or cross-selling** to keep values higher.  
- **Regional Opportunities:**  
  - **Brăila** → highest sales volume, ensure supply & strong retention.  
  - **Ploiești** → leverage the **highest order size behavior** and replicate in other regions.  
- **Stabilize Declines:** Address the **Sep 2024 low ($309)** → investigate if caused by **discount-heavy campaigns, economic slowdown, or competitive pressure**.


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
    3. What is the total profit?
  </span>
</div>
<a id="import-data"></a>


In [59]:


profit_summary = df.groupby('Region_Name').agg({'Profit':'sum'}).reset_index().sort_values(by='Profit', ascending=False)

fig = px.bar(
    profit_summary.round(2),
    x='Region_Name',
    y='Profit',
    text='Profit', 
    labels={'Profit':'Total Profit ($)', 'Region_Name':'Region'},
    color_discrete_sequence=['#34ebd8']
)

fig.update_layout(
    title='Total Profit per Region',
    title_x=0.5,
    template='plotly_dark'
)

profit_summary

Region_Name        Profit
0       Brăila  4.310655e+06
1      Craiova  3.391638e+06
4  Târgu Mureș  3.131380e+06
2       Galați  2.759370e+06
3     Ploiești  2.622160e+06

In [60]:
fig.show()


In [61]:

df['Order_Date'] = pd.to_datetime(df['Order_Date'])

monthly_summary = df.groupby(df['Order_Date'].dt.to_period('M')).agg({
    'Profit':'sum'
}).reset_index()

monthly_summary['Period'] = monthly_summary['Order_Date'].dt.strftime('%b %Y')


In [62]:
monthly_summary

Order_Date        Profit    Period
0     2022-10  149559.63902  Oct 2022
1     2022-11  672234.19450  Nov 2022
2     2022-12  907684.71200  Dec 2022
3     2023-01  630532.53186  Jan 2023
4     2023-02  527501.16078  Feb 2023
5     2023-03  688422.67450  Mar 2023
6     2023-04  548356.70956  Apr 2023
7     2023-05  719155.02406  May 2023
8     2023-06  696562.24858  Jun 2023
9     2023-07  723115.78812  Jul 2023
10    2023-08  714188.40262  Aug 2023
11    2023-09  761427.28596  Sep 2023
12    2023-10  579191.03572  Oct 2023
13    2023-11  679534.66306  Nov 2023
14    2023-12  790779.15092  Dec 2023
15    2024-01  581618.66950  Jan 2024
16    2024-02  679804.77290  Feb 2024
17    2024-03  484541.00618  Mar 2024
18    2024-04  710798.73612  Apr 2024
19    2024-05  705396.57826  May 2024
20    2024-06  786507.45382  Jun 2024
21    2024-07  668166.27372  Jul 2024
22    2024-08  676685.34450  Aug 2024
23    2024-09  563178.94446  Sep 2024
24    2024-10  570260.62280  Oct 2024

In [63]:
fig_profit = px.line(
    monthly_summary.round(1),
    x='Period',
    y='Profit',
    markers=True,
    text='Profit',
    labels={'Profit':'Total Profit ($)', 'Period':'Month'}
)

fig_profit.update_layout(
    title='Monthly Total Profit Trend',
    title_x=0.5,
    template='plotly_dark',
    xaxis_tickangle=-45
)

fig_profit.show()


# Profit Insights (Regional & Monthly)

---

## 1. Regional Profit Distribution
| Region       | Profit (USD) |
|--------------|-------------:|
| Brăila       | 4.31M |
| Craiova      | 3.39M |
| Galați       | 2.76M |
| Ploiești     | 2.62M |
| Târgu Mureș  | 3.13M |

### Insights:
- **Brăila** leads with the **highest profit (4.31M)** — consistent with its leadership in sales.  
- **Craiova (3.39M)** and **Târgu Mureș (3.13M)** form the second tier of profitable regions.  
- **Ploiești (2.62M)** shows strong **average order sizes**, but overall profit lags behind Brăila.  
- The regional spread suggests **profitability correlates strongly with sales volume**, but regions with higher order values (e.g., Ploiești) should be explored for **margin-focused strategies**.

---

## 2. Monthly Profit Trends
| Period   | Profit (USD) |
|----------|--------------:|
| Oct 2022 | 149.6K |
| Nov 2022 | 672.2K |
| Dec 2022 | 907.7K |
| Jan 2023 | 630.5K |
| Feb 2023 | 527.5K |
| Mar 2023 | 688.4K |
| Apr 2023 | 548.4K |
| May 2023 | 719.2K |
| Jun 2023 | 696.6K |
| Jul 2023 | 723.1K |
| Aug 2023 | 714.2K |
| Sep 2023 | 761.4K |
| Oct 2023 | 579.2K |
| Nov 2023 | 679.5K |
| Dec 2023 | 790.8K |
| Jan 2024 | 581.6K |
| Feb 2024 | 679.8K |
| Mar 2024 | 484.5K |
| Apr 2024 | 710.8K |
| May 2024 | 705.4K |
| Jun 2024 | 786.5K |
| Jul 2024 | 668.2K |
| Aug 2024 | 676.7K |
| Sep 2024 | 563.2K |
| Oct 2024 | 570.3K |

### Observations:
- **Profit followed sales trends closely**, peaking during **Nov–Dec (holiday period)** each year.  
- **Dec 2022:** Highest early profit (**907.7K**).  
- **Dec 2023:** Also strong (**790.8K**) → confirms **end-of-year demand**.  
- **Q1 Profit Weakness:**  
  - Jan–Mar in both years show dips, lowest being **Mar 2024 (484.5K)**.  
- **Mid-Year Stability (Apr–Jun 2024):** Profits stabilized around **705–787K**, indicating efficient sales execution.  
- **Q3 2024 Concern:** Profits dropped to **563K (Sep 2024)** → aligns with falling sales and average order size.  

---

## 3. Actionable Insights
- **Leverage Seasonal Peaks:** Focus campaigns on **Nov–Dec**, where profits surge every year.  
- **Address Q1 Dip:** Introduce **new-year promotions, retention campaigns, or upselling** to avoid profit slumps.  
- **Optimize Q3 2024:** Investigate why **profits dropped alongside sales and order sizes** — possible discount pressure, reduced demand, or competitive activity.  
- **Regional Strategy:**  
  - **Brăila** → invest to maintain dominance.  
  - **Ploiești** → already strong in order size, can push for **margin optimization**.  
  - **Lagging regions** → replicate best practices from high performers to lift profitability.  


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
   4. What is the sales distribution by time of day (Morning – Afternoon – Evening – Night)?
  </span>
</div>
<a id="import-data"></a>


In [64]:
def time_of_day(hour):
    if 6 <= hour < 12:
        return 'Morning'
    elif 12 <= hour < 18:
        return 'Afternoon'
    elif 18 <= hour < 24:
        return 'Evening'
    else:
        return 'Night'

df['Hour'] = pd.to_datetime(df['Time'], format='%H:%M').dt.hour
df['Time_of_Day'] = df['Hour'].apply(time_of_day)


In [65]:
sales_by_time = df.groupby('Time_of_Day').agg({'Total_Sales':'sum'}).reset_index()

order = ['Morning', 'Afternoon', 'Evening', 'Night']
sales_by_time['Time_of_Day'] = pd.Categorical(sales_by_time['Time_of_Day'], categories=order, ordered=True)
sales_by_time = sales_by_time.sort_values('Time_of_Day')


In [66]:
sales_by_time.round(2)

Time_of_Day  Total_Sales
2     Morning  19245756.97
0   Afternoon  19626492.03
1     Evening  18540014.11
3       Night  18962782.26

In [67]:

df['Hour'] = pd.to_datetime(df['Time'], format='%H:%M', errors='coerce').dt.hour

def get_time_of_day(hour):
    if pd.isna(hour):
        return "Unknown"
    if 5 <= hour < 12:
        return "Morning"
    elif 12 <= hour < 17:
        return "Afternoon"
    elif 17 <= hour < 21:
        return "Evening"
    else:
        return "Night"

df['Time_of_Day'] = df['Hour'].apply(get_time_of_day)

time_sales = df.groupby('Time_of_Day')['Total_Sales'].sum().reset_index()

order = ['Morning', 'Afternoon', 'Evening', 'Night', 'Unknown']
time_sales['Time_of_Day'] = pd.Categorical(time_sales['Time_of_Day'], categories=order, ordered=True)
time_sales = time_sales.sort_values('Time_of_Day').sort_values('Total_Sales', ascending=False)

fig = make_subplots(
    rows=1, cols=2,
    specs=[[{"type": "bar"}, {"type": "pie"}]],
    subplot_titles=("Sales by Time of Day (Bar)", "Sales Distribution by Time of Day (Pie)")
)

fig.add_trace(
    go.Bar(
        x=time_sales['Time_of_Day'],
        y=time_sales['Total_Sales'],
        text=[f"${v:,.0f}" for v in time_sales['Total_Sales']],
        textposition='outside',
        marker_color='#34c0eb',
        name="Sales"
    ),
    row=1, col=1
)

# Pie chart
fig.add_trace(
    go.Pie(
        labels=time_sales['Time_of_Day'],
        values=time_sales['Total_Sales'],
        hole=0.4,
        textinfo='label+percent',
        marker=dict(colors=['#34c0eb', '#ffab40', '#ff5252', '#4caf50', '#9c27b0'])
    ),
    row=1, col=2
)

fig.update_layout(
    title_text="Sales Distribution by Time of Day",
    title_x=0.5,
    template="plotly_dark",
    showlegend=True,

)

fig.show()


# Time-of-Day Sales Insights

---

## 1. Sales by Time of Day
| Time of Day | Total Sales (USD) |
|-------------|------------------:|
| Afternoon   | 19.63M |
| Morning     | 19.25M |
| Night       | 18.96M |
| Evening     | 18.54M |

---

## 2. Observations
- **Afternoon** leads with the **highest sales (19.63M)**, slightly above **Morning (19.25M)**.  
- **Evening** shows the **lowest sales (18.54M)** but still close to other periods.  
- The distribution is **fairly balanced across the day**, with no extreme gaps — only about **1.1M difference between highest and lowest periods**.  
- Suggests that **demand is consistent throughout the day**, though **Afternoons perform best**.

---

## 3. Actionable Insights
- **Strengthen Afternoon Campaigns:** Since afternoons generate the most sales, double down on **ads, promotions, and customer engagement** during this time.  
- **Boost Evenings:** Evenings are slightly weaker → explore **after-work promotions, flash sales, or social media ads** to lift performance.  
- **Maintain Consistency:** Since sales are evenly distributed, ensure **stock availability and customer support coverage** across all times.  


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
  What are the main KPIs, such as:
   - Total Sales
   - Total Quantity Sold
   - Total Profit
   - Average Order Size
   - Sales Growth Rate
   - Return Rate
  </span>
</div>
<a id="import-data"></a>


In [68]:

total_sales = df['Total_Sales'].sum()
total_quantity = df['Quantity_Sold'].sum()
total_profit = df['Profit'].sum()
average_order_size = (df['Total_Sales'].sum() / df['Quantity_Sold'].sum())

df['Order_Date'] = pd.to_datetime(df['Order_Date'])
df['Period'] = df['Order_Date'].dt.to_period('M')

monthly_sales = df.groupby('Period')['Total_Sales'].sum().sort_index()
sales_growth_rate = monthly_sales.pct_change().mean() * 100     


if 'Return_Flag' in df.columns:
    return_rate = (df['Return_Flag'].sum() / len(df)) * 100
else:
    return_rate = None


In [69]:
kpis = {
    "Total Sales ($)": total_sales,
    "Total Quantity Sold": total_quantity,
    "Total Profit ($)": total_profit,
    "Average Order Size ($)": average_order_size,
    "Sales Growth Rate (%)": sales_growth_rate,
    "Return Rate (%)": return_rate
}

kpis_df = pd.DataFrame(kpis.items(), columns=["KPI", "Value"])
kpis_df


KPI         Value
0         Total Sales ($)  7.637505e+07
1     Total Quantity Sold  2.172710e+05
2        Total Profit ($)  1.621520e+07
3  Average Order Size ($)  3.515197e+02
4   Sales Growth Rate (%)  1.721752e+01
5         Return Rate (%)  1.930048e+01

In [70]:

fig = go.Figure()

# Total Sales
fig.add_trace(go.Indicator(
    mode="number",
    value=total_sales,
    number={"prefix": "$", "font": {"size": 30, "color": "royalblue"}},
    title={"text": "<b>Total Sales</b>", "font": {"size": 18}},
    domain={'row': 0, 'column': 0}
))

# Total Quantity
fig.add_trace(go.Indicator(
    mode="number",
    value=total_quantity,
    number={"font": {"size": 30, "color": "darkorange"}},
    title={"text": "<b>Total Quantity Sold</b>", "font": {"size": 18}},
    domain={'row': 0, 'column': 1}
))

# Total Profit
fig.add_trace(go.Indicator(
    mode="number",
    value=total_profit,
    number={"prefix": "$", "font": {"size": 30, "color": "green"}},
    title={"text": "<b>Total Profit</b>", "font": {"size": 18}},
    domain={'row': 0, 'column': 2}
))

# Average Order Size
fig.add_trace(go.Indicator(
    mode="number",
    value=average_order_size,
    number={"prefix": "$", "font": {"size": 30, "color": "purple"}},
    title={"text": "<b>Average Order Size</b>", "font": {"size": 18}},
    domain={'row': 1, 'column': 0}
))

# Sales Growth Rate
fig.add_trace(go.Indicator(
    mode="number+delta",
    value=sales_growth_rate,
    number={"suffix": "%", "font": {"size": 30, "color": "crimson"}},
    delta={'reference': 0, 'relative': False, 'position': "right"},
    title={"text": "<b>Sales Growth Rate</b>", "font": {"size": 18}},
    domain={'row': 1, 'column': 1}
))

# Return Rate
fig.add_trace(go.Indicator(
    mode="number+delta",
    value=return_rate if return_rate is not None else 0,
    number={"suffix": "%", "font": {"size": 30, "color": "firebrick"}},
    delta={'reference': 5, 'relative': False, 'increasing': {'color': "red"}, 'decreasing': {'color': "green"}},
    title={"text": "<b>Return Rate</b>", "font": {"size": 18}},
    domain={'row': 1, 'column': 2}
))

# Layout
fig.update_layout(
    grid={'rows': 2, 'columns': 3, 'pattern': "independent"},
    title={"text": "📊 Main Business KPIs", "x": 0.5, "xanchor": "center", "font": {"size": 26, "color": "white"}},
    template="plotly_dark",
    height=600
)

fig.show()


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
    6. What are the top products by total sales?
  </span>
</div>
<a id="import-data"></a>


In [71]:

product_sales = (
    df.groupby('Product_Name', as_index=False)['Total_Sales']
    .sum().round(0)
    .sort_values(by='Total_Sales', ascending=False)
)

top_products = product_sales


base_hex = "#34ebd8"


def hex_to_rgb(hex_color):
    h = hex_color.lstrip('#')
    return tuple(int(h[i:i+2], 16) for i in (0, 2, 4))

def rgb_to_hex(rgb):
    return '#%02x%02x%02x' % rgb

def generate_gradient_colors(base_hex, n):
    base_rgb = hex_to_rgb(base_hex)
    gradient = []
    for i in range(n):
        factor = 0.5 + 0.5 * (i / max(n-1, 1))  
        r = int(base_rgb[0] * factor)
        g = int(base_rgb[1] * factor)
        b = int(base_rgb[2] * factor)
        gradient.append(rgb_to_hex((r,g,b)))
    return gradient[::-1]  

bar_colors = generate_gradient_colors(base_hex, len(top_products))

fig = make_subplots(
    rows=1, cols=2,
    specs=[[{"type":"bar"}, {"type":"pie"}]],
    subplot_titles=("Top Products (Bar)", "Top Products (Pie)")
)

fig.add_trace(
    go.Bar(
        x=top_products['Total_Sales'],
        y=top_products['Product_Name'],
        orientation='h',
        text=[f"${v:,.0f}" for v in top_products['Total_Sales']],
        textposition='outside',
        marker=dict(
            color=bar_colors,
            line=dict(color='black', width=1.2)
        )
    ),
    row=1, col=1
)

fig.add_trace(
    go.Pie(
        labels=top_products['Product_Name'],
        values=top_products['Total_Sales'],
        hole=0.4,
        textinfo='label+percent',
        marker=dict(colors=bar_colors)
    ),
    row=1, col=2
)

fig.update_layout(
    title_text="Top Products by Total Sales",
    title_x=0.5,
    template="plotly_dark",

)

fig.update_yaxes(autorange="reversed", row=1, col=1)
fig.update_xaxes(showgrid=True, gridcolor="lightgray", row=1, col=1)

fig.show()


# Product Sales Insights

---

## 1. Top  Products by Total Sales
| Product Name        | Total Sales (USD) |
|---------------------|------------------:|
| Televisions         | 16.31M |
| Treadmill           | 15.15M |
| Air Fryer           | 15.08M |
| Earbuds             | 14.96M |
| Camera Accessories  | 14.87M |

---

## 2. Observations
- **Televisions** lead all categories with **$16.3M sales**, slightly ahead of others.  
- **Treadmill, Air Fryer, Earbuds, and Camera Accessories** all cluster around **$14.8M–$15.2M**, showing **balanced product performance**.  
- The  products are **very close in sales volume**, indicating a **diverse and resilient portfolio** rather than dependence on a single hero product.  

---

## 3. Actionable Insights
- **Maintain TV Dominance:** Televisions are the **highest revenue driver** → focus on **premium models and bundles**.  
- **Leverage Balanced Categories:** Since other products are close in sales, cross-promote them (e.g., **Air Fryer + Earbuds seasonal bundles**).  
- **Explore Margins:** Investigate profitability per product → some categories (like accessories) may deliver **higher margins** even with slightly lower sales.  
- **Product Expansion:** The even spread suggests **room to scale new categories** without risking over-dependence on one product line.  


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
    7. What are the top regions by total sales?
  </span>
</div>
<a id="import-data"></a>


In [72]:

region_sales = (
    df.groupby('Region_Name', as_index=False)['Total_Sales']
    .sum().sort_values(by='Total_Sales', ascending=False)
)
top_regions = region_sales


base_hex = "#34ebd8"

def hex_to_rgb(hex_color):
    h = hex_color.lstrip('#')
    return tuple(int(h[i:i+2], 16) for i in (0, 2, 4))

def rgb_to_hex(rgb):
    return '#%02x%02x%02x' % rgb

def generate_gradient_colors(base_hex, n):
    base_rgb = hex_to_rgb(base_hex)
    gradient = []
    for i in range(n):
        factor = 0.5 + 0.5 * (i / max(n-1,1)) 
        r = int(base_rgb[0] * factor)
        g = int(base_rgb[1] * factor)
        b = int(base_rgb[2] * factor)
        gradient.append(rgb_to_hex((r,g,b)))
    return gradient[::-1]  

bar_colors = generate_gradient_colors(base_hex, len(top_regions))

fig = make_subplots(
    rows=1, cols=2,
    specs=[[{"type":"bar"}, {"type":"pie"}]],
    subplot_titles=("Top Regions (Bar)", "Top Regions (Pie)")
)
fig.add_trace(
    go.Bar(
        x=top_regions['Total_Sales'],
        y=top_regions['Region_Name'],
        orientation='h',
        text=[f"${v:,.0f}" for v in top_regions['Total_Sales']],
        textposition='outside',
        marker=dict(
            color=bar_colors,
            line=dict(color='black', width=1.2)
        )
    ),
    row=1, col=1
)

fig.add_trace(
    go.Pie(
        labels=top_regions['Region_Name'],
        values=top_regions['Total_Sales'],
        hole=0.4,
        textinfo='label+percent',
        marker=dict(colors=bar_colors)
    ),
    row=1, col=2
)

fig.update_layout(
    title_text="Top Regions by Total Sales",
    title_x=0.5,
    template="plotly_dark",

)

fig.update_yaxes(autorange="reversed", row=1, col=1)
fig.update_xaxes(showgrid=True, gridcolor="lightgray", row=1, col=1)

fig.show()


# Regional Sales Insights

---

## 1. Total Sales by Region
| Region       | Total Sales (USD) |
|--------------|------------------:|
| Brăila       | 20.55M |
| Craiova      | 15.84M |
| Târgu Mureș  | 14.75M |
| Galați       | 12.78M |
| Ploiești     | 12.46M |

---

## 2. Observations
- **Brăila** is the clear leader with **$20.6M**, significantly ahead of the other regions.  
- **Craiova (15.8M)** and **Târgu Mureș (14.8M)** form the middle tier.  
- **Galați (12.8M)** and **Ploiești (12.5M)** are at the lower end but still contribute strongly.  
- The gap between **Brăila and Ploiești** is around **$8M**, showing high **regional performance variance**.  



<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
    
   8. Who are the top customers by total sales?

  </span>
</div>
<a id="import-data"></a>


In [73]:

customer_sales = (
    df.groupby('Customer_Name', as_index=False)['Total_Sales']
    .sum()
    .sort_values(by='Total_Sales', ascending=False)
)
top_customers = customer_sales.head(10)
base_hex = "#34ebb7"

def hex_to_rgb(hex_color):
    h = hex_color.lstrip('#')
    return tuple(int(h[i:i+2], 16) for i in (0, 2, 4))

def rgb_to_hex(rgb):
    return '#%02x%02x%02x' % rgb

def generate_gradient_colors(base_hex, n):
    base_rgb = hex_to_rgb(base_hex)
    gradient = []
    for i in range(n):
        factor = 0.5 + 0.5 * (i / max(n-1,1))
        r = int(base_rgb[0] * factor)
        g = int(base_rgb[1] * factor)
        b = int(base_rgb[2] * factor)
        gradient.append(rgb_to_hex((r,g,b)))
    return gradient[::-1] 

bar_colors = generate_gradient_colors(base_hex, len(top_customers))

fig = make_subplots(
    rows=1, cols=2,
    specs=[[{"type":"bar"}, {"type":"pie"}]],
    subplot_titles=("Top Customers (Bar)", "Top Customers (Pie)")
)

fig.add_trace(
    go.Bar(
        x=top_customers['Total_Sales'],
        y=top_customers['Customer_Name'],
        orientation='h',
        text=[f"${v:,.0f}" for v in top_customers['Total_Sales']],
        textposition='outside',
        marker=dict(
            color=bar_colors,
            line=dict(color='black', width=1.2)
        )
    ),
    row=1, col=1
)

fig.add_trace(
    go.Pie(
        labels=top_customers['Customer_Name'],
        values=top_customers['Total_Sales'],
        hole=0.4,
        textinfo='label+percent',
        marker=dict(colors=bar_colors)
    ),
    row=1, col=2
)

fig.update_layout(
    title_text="Top 10 Customers by Total Sales",
    title_x=0.5,
    template="plotly_dark",

)

fig.update_yaxes(autorange="reversed", row=1, col=1)
fig.update_xaxes(showgrid=True, gridcolor="lightgray", row=1, col=1)

fig.show()


In [74]:
customer_sales

Customer_Name  Total_Sales
26        Ioan Popescu  2296362.904
7      Andreea Ionescu  2282306.642
28       Ioana Ionescu  2238625.136
6    Andreea Georgescu  1580785.624
3        Ana Georgescu  1580785.624
40    Nicoleta Ionescu  1580785.624
32       Laura Ionescu  1566729.362
1   Alexandru Petrescu  1566729.362
39     Nicolae Popescu  1566729.362
10    Ciprian Petrescu  1566729.362
2      Alina Georgescu  1544845.434
34   Mihaela Georgescu  1544845.434
20       Diana Ionescu  1544845.434
22     Dumitru Popescu  1544845.434
30       Ionuț Popescu  1544845.434
41      Oana Georgescu  1544845.434
38  Mădălina Georgescu  1541910.282
44      Roxana Ionescu  1541910.282
47      Sorin Petrescu  1541910.282
36     Mircea Petrescu  1541910.282
52     Ștefan Petrescu  1530789.172
31       Iulia Ionescu  1530789.172
25    Gheorghe Popescu  1530789.172
49      Vasile Popescu  1527854.020
42        Paul Popescu  1527854.020
4    Ana Maria Ionescu  1527854.020
15    Cristian Popescu  1527854.020
24   Georgiana Ionescu  1527854.020
43      Raluca Ionescu  1523047.856
17   Daniela Georgescu  1523047.856
14     Cosmin Petrescu  1523047.856
16  Cristina Georgescu  1508991.594
13    Corina Georgescu  1508991.594
21     Dragoș Petrescu  1508991.594
37    Monica Georgescu  1505970.092
27     Ioana Georgescu  1503034.940
0      Adrian Petrescu  1487107.666
50     Virgil Petrescu  1487107.666
48   Valentin Petrescu  1487107.666
12  Constantin Popescu  1487107.666
33      Marius Popescu  1487107.666
45  Ruxandra Georgescu  1484172.514
9      Camelia Ionescu  1484172.514
23    Gabriel Petrescu  1465310.088
18     Darius Petrescu   790392.812
46      Răzvan Popescu   790392.812
35       Mihai Popescu   776336.550
51       Vlad Petrescu   776336.550
5    Ana-Maria Ionescu   776336.550
19       Denis Popescu   754452.622
11    Claudiu Petrescu   754452.622
8     Bianca Georgescu   754452.622
29         Ion Popescu   732655.044

# Customer Sales Insights

---

## 1. Top Customers by Total Sales
| Customer Name        | Total Sales (USD) |
|----------------------|------------------:|
| Ioan Popescu         | 2.30M |
| Andreea Ionescu      | 2.28M |
| Ioana Ionescu        | 2.24M |
| Andreea Georgescu    | 1.58M |
| Ana Georgescu        | 1.58M |
| Nicoleta Ionescu     | 1.58M |
| Laura Ionescu        | 1.57M |
| Alexandru Petrescu   | 1.57M |
| Nicolae Popescu      | 1.57M |
| Ciprian Petrescu     | 1.57M |
| ...                  | ... |

---

## 2. Observations
- **Top 3 customers (Ioan Popescu, Andreea Ionescu, Ioana Ionescu)** together contribute **~$6.76M**, a significant portion of total customer sales.  
- There is a **cluster of customers between $1.5M – $1.6M**, indicating a strong **upper-mid tier** of loyal buyers.  
- A few customers like **Darius Petrescu, Răzvan Popescu, Mihai Popescu, Vlad Petrescu** contribute below **$0.8M**, marking them as **lower-value but potential growth targets**.  
- Common family names (Popescu, Ionescu, Georgescu, Petrescu) dominate the list → indicates **strong household or regional loyalty patterns**.  

---

## 3. Actionable Insights
- **VIP Retention Strategy:** Offer **exclusive loyalty perks** to the top 3–5 customers to protect high-value accounts.  
- **Mid-Tier Engagement:** Customers in the **$1.5M–$1.6M range** are highly engaged — push **cross-sell/up-sell** campaigns.  
- **Growth Opportunities:** Target the **sub-$1M customers** with personalized promotions to lift them into higher contribution tiers.  
- **Customer Clustering:** Since many top customers share last names, design **family/household-level bundles** to maximize collective spending.  


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
    
   
 9. Which days of the week have the highest sales?
  </span>
</div>
<a id="import-data"></a>


In [75]:

day_sales = (
    df.groupby('Day_of_Week', as_index=False)['Total_Sales']
    .sum().sort_values(by='Total_Sales')
)

week_order = ['Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday']
day_sales['Day_of_Week'] = pd.Categorical(day_sales['Day_of_Week'], categories=week_order, ordered=True)
day_sales = day_sales.sort_values('Day_of_Week').sort_values('Total_Sales', ascending=False)

base_hex = "#34ebd8"

def hex_to_rgb(hex_color):
    h = hex_color.lstrip('#')
    return tuple(int(h[i:i+2], 16) for i in (0, 2, 4))

def rgb_to_hex(rgb):
    return '#%02x%02x%02x' % rgb

def generate_gradient_colors(base_hex, n):
    base_rgb = hex_to_rgb(base_hex)
    gradient = []
    for i in range(n):
        factor = 0.5 + 0.5*(i / max(n-1,1))
        r = int(base_rgb[0]*factor)
        g = int(base_rgb[1]*factor)
        b = int(base_rgb[2]*factor)
        gradient.append(rgb_to_hex((r,g,b)))
    return gradient[::-1]

colors = generate_gradient_colors(base_hex, len(day_sales))

fig = make_subplots(
    rows=1, cols=2,
    specs=[[{"type":"bar"}, {"type":"pie"}]],
    subplot_titles=("Sales by Day of Week (Bar)", "Sales Distribution by Day (Pie)")
)

fig.add_trace(
    go.Bar(
        x=day_sales['Day_of_Week'],
        y=day_sales['Total_Sales'],
        text=[f"${v:,.0f}" for v in day_sales['Total_Sales']],
        textposition='outside',
        marker=dict(color=colors, line=dict(color='black', width=1.2))
    ),
    row=1, col=1
)

fig.add_trace(
    go.Pie(
        labels=day_sales['Day_of_Week'],
        values=day_sales['Total_Sales'],
        hole=0.4,
        textinfo='label+percent',
        marker=dict(colors=colors)
    ),
    row=1, col=2
)

fig.update_layout(
    title_text="Sales by Day of Week",
    title_x=0.5,
    template="plotly_dark",

)

fig.show()


# Day of Week Sales Insights

---

## 1. Total Sales by Day of the Week
| Day        | Total Sales (USD) |
|------------|------------------:|
| Thursday   | 12.00M |
| Wednesday  | 11.61M |
| Saturday   | 11.42M |
| Friday     | 11.34M |
| Tuesday    | 10.78M |
| Sunday     | 10.02M |
| Monday     | 9.21M |

---

## 2. Observations
- **Thursday is the strongest sales day** with ~$12M, slightly ahead of Wednesday and Saturday.  
- **Weekdays (Tue–Thu)** outperform weekends, suggesting stronger B2B or weekday-driven demand.  
- **Monday has the lowest sales ($9.2M)**, indicating a slow start to the week.  
- **Weekend sales (Saturday & Sunday)** remain strong, making up ~$21.4M combined.  

---

## 3. Actionable Insights
- **Promotions on Monday:** Boost weaker Monday sales with **special deals or discounts** to encourage early-week purchases.  
- **Maximize Thursday demand:** Since Thursday is the peak, consider **flash sales or product launches** mid-to-late week.  
- **Weekend Bundles:** Leverage consumer shopping patterns on **Saturday & Sunday** with **family/entertainment-focused offers**.  
- **Staffing & Logistics:** Allocate **more staff and inventory** midweek (Wed–Thu) to handle higher volumes.  


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
     10. What is the ranking of sales reps by total sales and quantity sold?
  </span>
</div>
<a id="import-data"></a>


In [76]:
rep_sales = df.groupby('Sales_Rep_Name', as_index=False).agg({
    'Total_Sales':'sum',
    'Quantity_Sold':'sum'
}).sort_values(by='Total_Sales', ascending=True)

top_reps = rep_sales

base_hex = "#34c0eb"
def hex_to_rgb(hex_color):
    h = hex_color.lstrip('#')
    return tuple(int(h[i:i+2], 16) for i in (0, 2, 4))
def rgb_to_hex(rgb):
    return '#%02x%02x%02x' % rgb
def generate_gradient_colors(base_hex, n):
    base_rgb = hex_to_rgb(base_hex)
    gradient = []
    for i in range(n):
        factor = 0.5 + 0.5*(i / max(n-1,1))
        r = int(base_rgb[0]*factor)
        g = int(base_rgb[1]*factor)
        b = int(base_rgb[2]*factor)
        gradient.append(rgb_to_hex((r,g,b)))
    return gradient[::-1]

colors_sales = generate_gradient_colors(base_hex, len(top_reps))
colors_qty = ['#ffab40', '#ff6f61', '#ffa07a', '#ffd700', '#90ee90'][:len(top_reps)]

fig = make_subplots(
    rows=2, cols=2,
    specs=[[{"type":"bar"}, {"type":"pie"}],
           [{"type":"bar"}, {"type":"pie"}]],
    subplot_titles=("Total Sales (Bar)", "Total Sales (Pie)",
                    "Quantity Sold (Bar)", "Quantity Sold (Pie)")
)

fig.add_trace(
    go.Bar(
        x=top_reps['Total_Sales'],
        y=top_reps['Sales_Rep_Name'],
        orientation='h',
        text=[f"${v:,.0f}" for v in top_reps['Total_Sales']],
        textposition='outside',
        marker=dict(color=colors_sales, line=dict(color='black', width=1.2))
    ),
    row=1, col=1
)

fig.add_trace(
    go.Pie(
        labels=top_reps['Sales_Rep_Name'],
        values=top_reps['Total_Sales'],
        hole=0.4,
        textinfo='label+percent',
        marker=dict(colors=colors_sales)
    ),
    row=1, col=2
)

fig.add_trace(
    go.Bar(
        x=top_reps['Quantity_Sold'],
        y=top_reps['Sales_Rep_Name'],
        orientation='h',
        text=top_reps['Quantity_Sold'],
        textposition='outside',
        # marker=dict(color= colors_qty, line=dict(color='black', width=1.2))
    ),
    row=2, col=1
)

fig.add_trace(
    go.Pie(
        labels=top_reps['Sales_Rep_Name'],
        values=top_reps['Quantity_Sold'],
        hole=0.4,
        textinfo='label+percent',
        # marker=dict(colors=colors_qty)
    ),
    row=2, col=2
)

fig.update_layout(
    title_text="Sales Reps Overview: Total Sales & Quantity Sold",
    title_x=0.5,
    template="plotly_dark",

)

fig.show()


# Sales Representative Performance Insights

---

## 1. Sales & Quantity by Representative
| Sales Rep          | Total Sales (USD) | Quantity Sold |
|--------------------|------------------:|--------------:|
| Cristian Popescu   | 17.18M | 47,095 |
| Iulia Ionescu      | 15.66M | 43,512 |
| Alina Georgescu    | 14.94M | 44,006 |
| Dragoș Petrescu    | 14.42M | 41,213 |
| Cosmin Petrescu    | 14.17M | 41,445 |

---

## 2. Observations
- **Cristian Popescu** leads with **$17.2M sales and ~47K units sold**, significantly ahead of peers.  
- **Iulia Ionescu** follows with **$15.7M**, maintaining strong balance between sales value and quantity.  
- **Alina Georgescu** is competitive with **highest quantity sold (44K)** after Cristian, but slightly lower in sales value.  
- **Cosmin & Dragoș Petrescu** perform steadily with **~$14.2M each**, but trail behind the leaders.  

---

## 3. Actionable Insights
- **Recognize Cristian Popescu:** Top performer — leverage his sales strategies for team training.  
- **Support Alina & Iulia:** Both have **strong quantity sales**, indicating potential for **upselling or premium product focus**.  
- **Boost Cosmin & Dragoș:** Provide **targeted coaching or product mix adjustments** to help close the gap with leaders.  
- **Balanced Targets:** Set performance benchmarks based on **Cristian’s sales and Iulia’s consistency** to drive overall team growth.  


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
      11. What is the product category filter?
  </span>
</div>
<a id="import-data"></a>


In [77]:

category_sales = df.groupby('Product_Category', as_index=False)['Total_Sales'].sum().sort_values(by='Total_Sales', ascending=True)

fig = make_subplots(
    rows=1, cols=2,
    specs=[[{"type":"bar"}, {"type":"pie"}]],
    subplot_titles=("Sales by Product Category (Bar)", "Sales Distribution by Category (Pie)")
)

fig.add_trace(
    go.Bar(
        x=category_sales['Total_Sales'],
        y=category_sales['Product_Category'],
        orientation='h',
        text=[f"${v:,.0f}" for v in category_sales['Total_Sales']],
        textposition='outside',
        marker=dict(color=['#34ebd8', '#0a9396', '#0077b6'][:len(category_sales)], line=dict(color='black', width=1.2))
    ),
    row=1, col=1
)

fig.add_trace(
    go.Pie(
        labels=category_sales['Product_Category'],
        values=category_sales['Total_Sales'],
        hole=0.4,
        textinfo='label+percent',
        marker=dict(colors=['#34ebd8', '#0a9396', '#0077b6'][:len(category_sales)])
    ),
    row=1, col=2
)

fig.update_layout(
    title_text="Product Category Sales Overview",
    title_x=0.5,
    template="plotly_dark",

)

fig.show()


# Product Category Sales Insights

---

## 1. Sales by Category
| Product Category     | Total Sales (USD) |
|----------------------|------------------:|
| Electronics          | 46.14M |
| Sports               | 15.15M |
| Kitchen Appliances   | 15.08M |

---

## 2. Observations
- **Electronics** dominate sales with **$46.1M**, accounting for nearly **60% of total sales**, making it the core revenue driver.  
- **Sports** and **Kitchen Appliances** are almost equal contributors, each around **$15M**, together making up ~40% of sales.  
- Sales distribution indicates a **heavy reliance on Electronics**, while the other two categories play a **supporting role**.  

---

## 3. Actionable Insights
- **Leverage Electronics Growth:** Expand product range, bundles, or premium lines since it’s the highest-performing category.  
- **Diversify Revenue:** Increase marketing and promotions for **Sports** and **Kitchen Appliances** to balance reliance on Electronics.  
- **Cross-Selling Potential:** Pair **Electronics with Kitchen Appliances** (e.g., smart kitchen gadgets) or **Sports products with wearables**, boosting complementary sales.  
- **Risk Management:** Monitor Electronics dependency — a dip in this category could significantly impact overall performance.  


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
      12. What is the region filter?
  </span>
</div>
<a id="import-data"></a>


In [78]:

colors = ['#34ebd8', '#0a9396', '#0077b6', '#f4d35e', '#ee6c4d']

region_sales = df.groupby('Region_Name', as_index=False)['Total_Sales'] \
                 .sum().sort_values(by='Total_Sales', ascending=False)

monthly_sales = df.groupby(['Region_Name', 'Month'], as_index=False)['Total_Sales'] \
                  .sum().sort_values('Month')

fig = make_subplots(
    rows=2, cols=2,
    specs=[[{"type":"bar"}, {"type":"pie"}],
           [{"type":"scatter", "colspan": 2}, None]],
    subplot_titles=("Sales by Region (Bar)", "Sales Distribution by Region (Pie)",
                    "Monthly Sales Trend (Line)", "")
)

fig.add_trace(
    go.Bar(
        x=region_sales['Total_Sales'],
        y=region_sales['Region_Name'],
        orientation='h',
        text=[f"${v:,.0f}" for v in region_sales['Total_Sales']],
        textposition='outside',
        marker=dict(color=colors[:len(region_sales)],
                    line=dict(color='black', width=1.2))
    ),
    row=1, col=1
)

fig.add_trace(
    go.Pie(
        labels=region_sales['Region_Name'],
        values=region_sales['Total_Sales'],
        hole=0.4,
        textinfo='label+percent',
        marker=dict(colors=colors[:len(region_sales)])
    ),
    row=1, col=2
)

for i, region in enumerate(region_sales['Region_Name']):
    region_month = monthly_sales[monthly_sales['Region_Name']==region]
    fig.add_trace(
        go.Scatter(
            x=region_month['Month'],
            y=region_month['Total_Sales'],
            mode='lines+markers',
            name=region,
            line=dict(color=colors[i], width=3),
            marker=dict(size=8),
            hovertemplate='%{x} - %{y:,.0f} $<extra></extra>'
        ),
        row=2, col=1
    )

fig.update_layout(
    title_text="Region Sales Overview with Monthly Trend",
    title_x=0.5,
    template="plotly_dark",

    yaxis=dict(autorange="reversed") 
)

fig.show()


In [79]:
region_sales

Region_Name   Total_Sales
0       Brăila  2.055021e+07
1      Craiova  1.584351e+07
4  Târgu Mureș  1.475039e+07
2       Galați  1.277580e+07
3     Ploiești  1.245514e+07

In [80]:
monthly_sales

Region_Name  Month  Total_Sales
0        Brăila      1  1511138.772
12      Craiova      1  1213690.632
24       Galați      1   880927.080
36     Ploiești      1  1207556.240
48  Târgu Mureș      1  1010532.974
1        Brăila      2  1682856.032
25       Galați      2   876845.992
13      Craiova      2  1271858.742
37     Ploiești      2  1045680.064
49  Târgu Mureș      2   962170.374
50  Târgu Mureș      3  1394850.116
38     Ploiești      3  1027799.124
26       Galați      3  1040972.526
14      Craiova      3  1071674.142
2        Brăila      3  1060311.824
27       Galați      4  1245576.706
39     Ploiești      4   632760.502
51  Târgu Mureș      4  1056636.284
15      Craiova      4  1042375.950
3        Brăila      4  2108768.376
4        Brăila      5  1495709.644
16      Craiova      5  1564244.220
52  Târgu Mureș      5  1539776.986
40     Ploiești      5  1154750.806
28       Galați      5   932846.508
17      Craiova      6  1774027.794
29       Galați      6  1075039.812
5        Brăila      6  2100548.736
41     Ploiești      6   997524.198
53  Târgu Mureș      6  1157855.402
30       Galați      7  1146488.398
18      Craiova      7  1111959.618
42     Ploiești      7   930587.548
54  Târgu Mureș      7  1304810.408
6        Brăila      7  1674190.804
7        Brăila      8  2024583.704
31       Galați      8  1125399.286
19      Craiova      8  1122437.316
55  Târgu Mureș      8  1254367.422
43     Ploiești      8  1025729.408
56  Târgu Mureș      9  1264583.342
44     Ploiești      9  1060753.012
32       Galați      9  1294123.028
20      Craiova      9  1458183.804
8        Brăila      9  1081474.680
9        Brăila     10  1546826.996
21      Craiova     10  1753304.322
57  Târgu Mureș     10   758118.240
45     Ploiești     10  1166593.890
33       Galați     10   913769.516
10       Brăila     11  1924183.976
22      Craiova     11  1131002.334
58  Târgu Mureș     11  1415068.358
46     Ploiești     11   918728.008
34       Galați     11  1007528.698
11       Brăila     12  2339619.568
23      Craiova     12  1328746.188
35       Galați     12  1236279.440
47     Ploiești     12  1286672.948
59  Târgu Mureș     12  1631624.544

# Regional Sales Performance Insights

---

## 1. Regional Totals
| Region      | Total Sales (USD) |
|-------------|------------------:|
| Brăila      | 20.55M |
| Craiova     | 15.84M |
| Târgu Mureș | 14.75M |
| Galați      | 12.78M |
| Ploiești    | 12.46M |

🔹 **Brăila leads with $20.6M (≈27% of total),** followed by Craiova ($15.8M).  
🔹 **Ploiești has the lowest performance** at $12.5M.  

---

## 2. Monthly Sales Trends (Highlights)
- **Brăila**
  - Strong growth in **Nov–Dec**: $1.9M → $2.34M (peak month).
  - Consistently above $1.5M except for Sep ($1.08M).
- **Craiova**
  - Steady performance, peaking in **Oct ($1.75M)**.
  - Relatively stable range: $1.1M–$1.7M.
- **Târgu Mureș**
  - Volatile pattern with **Dec peak at $1.63M**, but weak **Oct ($0.76M)**.
- **Galați**
  - Smaller market, with sales fluctuating between **$0.87M–$1.29M**.
  - Best month: **Sep ($1.29M)**.
- **Ploiești**
  - Lowest contribution overall.
  - Peaked at **Dec ($1.29M)**, weakest in **Apr ($0.63M)**.

---

## 3. Insights & Recommendations
- **Brăila is the growth engine**: Strong December momentum suggests holiday-driven demand — replicate marketing strategies region-wide.  
- **Craiova is stable**: Focus on scaling via cross-selling & premium product push.  
- **Târgu Mureș is volatile**: Needs consistent engagement to reduce fluctuations.  
- **Galați & Ploiești lag**: Consider targeted promotions, local partnerships, or sales rep support.  
- **Seasonality effect**: December consistently drives peaks → plan **stocking, promotions, and staffing** for Q4.  


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >
     13 📊 Sales Analysis: Hourly, Time of Day, Monthly & Yearly Trends
<a id="import-data"></a>


In [81]:

df['Hour'] = pd.to_datetime(df['Time'], format='%H:%M', errors='coerce').dt.hour
df['Year_Month'] = df['Year'].astype(str) + '-' + df['Month'].astype(str).str.zfill(2)

def get_time_of_day(hour):
    if pd.isna(hour):
        return "Unknown"
    if 5 <= hour < 12:
        return "Morning"
    elif 12 <= hour < 17:
        return "Afternoon"
    elif 17 <= hour < 21:
        return "Evening"
    else:
        return "Night"

df['Time_of_Day'] = df['Hour'].apply(get_time_of_day)

hour_sales = df.groupby('Hour', as_index=False)['Total_Sales'].sum()
time_sales = df.groupby('Time_of_Day', as_index=False)['Total_Sales'].sum()
time_order = ['Morning','Afternoon','Evening','Night','Unknown']
time_sales['Time_of_Day'] = pd.Categorical(time_sales['Time_of_Day'], categories=time_order, ordered=True)
time_sales = time_sales.sort_values('Time_of_Day')
monthly_sales = df.groupby(['Year','Month'], as_index=False)['Total_Sales'].sum()
monthly_sales['Year_Month'] = monthly_sales['Year'].astype(str) + '-' + monthly_sales['Month'].astype(str).str.zfill(2)
yearly_sales = df.groupby('Year', as_index=False)['Total_Sales'].sum()

fig = make_subplots(
    rows=2, cols=2,
    specs=[[{"type":"bar"}, {"type":"pie"}],
           [{"type":"scatter"}, {"type":"scatter"}]],
    subplot_titles=("Hourly Sales Distribution", "Sales by Time of Day",
                    "Monthly Sales Trend", "Yearly Sales Trend")
)

colors_hour = ['#34c0eb' if h < 12 else '#0a9396' if h < 17 else '#f4d35e' if h < 21 else '#ee6c4d' for h in hour_sales['Hour']]
fig.add_trace(
    go.Bar(
        x=hour_sales['Hour'],
        y=hour_sales['Total_Sales'],
        textposition='outside',
        marker_color=colors_hour,
        name='Hourly Sales',
        hovertemplate='Hour: %{x}<br>Sales: $%{y:,.0f}<extra></extra>'
    ),
    row=1, col=1
)
fig.update_xaxes(title_text="Hour of Day", row=1, col=1)
fig.update_yaxes(title_text="Total Sales ($)", row=1, col=1)

fig.add_trace(
    go.Pie(
        labels=time_sales['Time_of_Day'],
        values=time_sales['Total_Sales'],
        hole=0.4,
        textinfo='label+percent',
        marker=dict(colors=['#34c0eb','#0a9396','#0077b6','#f4d35e','#ee6c4d']),
        name='Time of Day Sales',
        hovertemplate='%{label}: $%{value:,.0f} (%{percent})<extra></extra>'
    ),
    row=1, col=2
)

for year in monthly_sales['Year'].unique():
    df_year = monthly_sales[monthly_sales['Year']==year]
    fig.add_trace(
        go.Scatter(
            x=df_year['Year_Month'],
            y=df_year['Total_Sales'],
            mode='lines+markers+text',
            textposition='top center',
            name=str(year),
            line=dict(width=3),
            marker=dict(size=8),
            hovertemplate='Month: %{x}<br>Sales: $%{y:,.0f}<extra></extra>'
        ),
        row=2, col=1
    )
fig.update_xaxes(title_text="Month", row=2, col=1)
fig.update_yaxes(title_text="Total Sales ($)", row=2, col=1)

fig.add_trace(
    go.Scatter(
        x=yearly_sales['Year'],
        y=yearly_sales['Total_Sales'],
        mode='lines+markers+text',
        textposition='top center',
        line=dict(color='#f4d35e', width=4),
        marker=dict(size=12, color='#ee6c4d'),
        name='Yearly Sales',
        hovertemplate='Year: %{x}<br>Sales: $%{y:,.0f}<extra></extra>'
    ),
    row=2, col=2
)
fig.update_xaxes(title_text="Year", row=2, col=2)
fig.update_yaxes(title_text="Total Sales ($)", row=2, col=2)

fig.update_layout(
    title_text="📊 Sales Analysis: Hourly, Time of Day, Monthly & Yearly Trends",
    title_x=0.5,
    template='plotly_dark',
    font=dict(size=14),
    showlegend=True
)

fig.show()


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >14. What is the difference between sales and returns by region?
<a id="import-data"></a>


In [82]:

region_sales_returns = df.groupby(['Region_Name', 'Return_Flag'], as_index=False)['Total_Sales'].sum()

region_pivot = region_sales_returns.pivot(index='Region_Name', columns='Return_Flag', values='Total_Sales').fillna(0)
region_pivot = region_pivot.rename(columns={0: 'Sales', 1: 'Returns'})

fig = go.Figure()

fig.add_trace(
    go.Bar(
        x=region_pivot.index,
        y=region_pivot['Sales'],
        name='Sales',
        marker_color='#34c0eb',
        textposition='auto'
    )
)

fig.add_trace(
    go.Bar(
        x=region_pivot.index,
        y=region_pivot['Returns'],
        name='Returns',
        marker_color='#ff6b6b',
        textposition='auto'
    )
)

fig.update_layout(
    title_text="Sales vs Returns by Region",
    title_x=0.5,
    template="plotly_dark",
    barmode='group', 
    xaxis_title="Region",
    yaxis_title="Amount ($)",

    font=dict(size=14)
)

fig.show()


# Regional Returns Analysis

---

## 1. Sales vs Returns by Region
| Region       | Sales (USD) | Returns (USD) | Return Rate (%) |
|--------------|------------:|--------------:|----------------:|
| Brăila       | 17.29M | 3.26M | **18.8%** |
| Craiova      | 12.62M | 3.22M | **25.5%** |
| Galați       | 10.12M | 2.66M | **26.3%** |
| Ploiești     | 10.02M | 2.44M | **24.4%** |
| Târgu Mureș  | 11.97M | 2.78M | **23.2%** |

---

## 2. Observations
- **Brăila** has the **lowest return rate (18.8%)**, indicating stronger product-market fit or better customer satisfaction.  
- **Galați** has the **highest return rate (26.3%)**, which significantly reduces net contribution.  
- **Craiova & Ploiești** both have **~25% return rates**, suggesting systemic issues (e.g., quality, mismatch, or logistics).  
- **Târgu Mureș** sits in the middle at **23.2%** — still relatively high.  

---

## 3. Actionable Insights
- **Best Practice Benchmark:** Analyze Brăila’s sales processes and product quality to replicate its lower return rate in other regions.  
- **Target Return Reduction:** Focus on **Galați and Craiova** — reducing returns by just 5% could recover ~$1M in profit annually.  
- **Customer Feedback Loops:** Implement post-return surveys or customer service analysis in high-return regions.  
- **Product Quality Checks:** Investigate product categories with the highest returns (e.g., fragile items in Electronics or size/fit issues in Sports).  
- **Incentivize Exchanges over Refunds:** Especially in **Craiova and Galați** to reduce lost revenue.  


<div style="text-align: center; margin: 30px 0;">
  <span style="
    font-family: 'Playfair Display', serif;
    background-color: hsla(176, 94%, 54%, 0.91);
    color: black;  
    font-size: 1.5em;
    font-weight: 600;
    font-style: italic;
    padding: 12px 24px;
    border-radius: 12px;
    display: inline-block;
    box-shadow: 0 4px 12px rgba(18, 17, 17, 0.15);
    transition: transform 0.2s ease, box-shadow 0.2s ease;
    cursor: default;
  " 
  onmouseover="this.style.transform='scale(1.03)'; this.style.boxShadow='0 6px 16px rgba(24, 22, 22, 0.2)'"
  onmouseout="this.style.transform='scale(1)'; this.style.boxShadow='0 4px 12px rgba(0, 0, 0, 0.15)'"
  >15.  What is the profit margin trend over time?
<a id="import-data"></a>


In [83]:
df['Order_Date'] = pd.to_datetime(df['Order_Date'])

df['YearMonth'] = df['Order_Date'].dt.to_period('M')

monthly_profit = df.groupby('YearMonth').agg({
    'Profit': 'sum',
    'Total_Sales': 'sum'
}).reset_index()

monthly_profit['Profit_Margin'] = monthly_profit['Profit'] / monthly_profit['Total_Sales'] * 100

fig = go.Figure()

fig.add_trace(
    go.Scatter(
        x=monthly_profit['YearMonth'].astype(str),
        y=monthly_profit['Profit_Margin'],
        mode='lines+markers',
        line=dict(color='#34c0eb', width=3),
        marker=dict(size=8),
        name='Profit Margin (%)',
        text=[f"{v:.2f}%" for v in monthly_profit['Profit_Margin']],
        textposition='top center'
    )
)

fig.update_layout(
    title_text="Profit Margin Trend Over Time",
    title_x=0.5,
    xaxis_title="Month",
    yaxis_title="Profit Margin (%)",
    template="plotly_dark",

    font=dict(size=14)
)

fig.show()


# Profit Margin Analysis

---

## 1. Profit Margin by Month
| Month      | Profit Margin (%) |
|------------|------------------:|
| Oct 2022   | 23.1% |
| Nov 2022   | 21.3% |
| Dec 2022   | 22.0% |
| Jan 2023   | 20.5% |
| Feb 2023   | 20.4% |
| Mar 2023   | 21.5% |
| Apr 2023   | 19.2% |
| May 2023   | 22.0% |
| Jun 2023   | 20.9% |
| Jul 2023   | **23.6%** (peak) |
| Aug 2023   | 21.7% |
| Sep 2023   | 21.3% |
| Oct 2023   | 21.1% |
| Nov 2023   | 21.0% |
| Dec 2023   | 21.4% |
| Jan 2024   | 21.2% |
| Feb 2024   | 20.9% |
| Mar 2024   | 20.2% |
| Apr 2024   | 22.0% |
| May 2024   | 20.6% |
| Jun 2024   | 20.9% |
| Jul 2024   | 21.5% |
| Aug 2024   | 20.7% |
| Sep 2024   | 21.8% |
| Oct 2024   | 20.8% |

---

## 2. Observations
- Profit margins are **generally stable between 20%–22%**, showing a healthy consistency.  
- **Lowest margin:** Apr 2023 (**19.2%**) — below average, potentially due to higher costs or discounts.  
- **Highest margin:** Jul 2023 (**23.6%**) — suggests strong profitability despite moderate sales.  
- **2024 margins** remain close to 21%, showing stability compared to **slight volatility in 2023**.  

---

## 3. Actionable Insights
- **Maintain Consistency:** Profit margins are steady; continue balancing cost controls with pricing strategy.  
- **Investigate Outliers:** Review **Apr 2023 (low)** and **Jul 2023 (high)** to identify causes — useful for forecasting and planning.  
- **Focus on Efficiency:** Even small improvements (1–2%) in profit margin can yield **significant impact** given sales scale (~$3M per month).  
- **Leverage High-Margin Months:** Strategies from **Jul 2023 & Apr 2024** could be applied to other periods.  
